commit 9a1c2cfb58a9ae7026fac04af4206c2a74566ba9 (refs/changes/79/5579/3)
Author: Ryan Zezeski <rpz@joyent.com>
Date:   2019-03-01T09:17:47-07:00 (7 months ago)
    
    OS-7564 panic in mac_hw_emul()
    Reviewed by: Robert Mustacchi <rm@joyent.com>
    Approved by: Jason King <jason.king@joyent.com>

diff --git a/usr/src/uts/common/io/mac/mac_util.c b/usr/src/uts/common/io/mac/mac_util.c
index 111b323f81..334d1d034b 100644
--- a/usr/src/uts/common/io/mac/mac_util.c
+++ b/usr/src/uts/common/io/mac/mac_util.c
@@ -20,7 +20,7 @@
  */
 /*
  * Copyright (c) 2008, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -1239,9 +1239,19 @@ mac_hw_emul(mblk_t **mp_chain, mblk_t **otail, uint_t *ocount, mac_emul_t emul)
 			 */
 			mac_sw_lso(mp, emul, &tmphead, &tmptail,
 			    &tmpcount);
+			if (tmphead == NULL) {
+				/* mac_sw_lso() freed the mp. */
+				mp = next;
+				continue;
+			}
 			count += tmpcount;
 		} else if ((flags & HCK_NEEDED) && (emul & MAC_HWCKSUM_EMULS)) {
 			tmp = mac_sw_cksum(mp, emul);
+			if (tmp == NULL) {
+				/* mac_sw_cksum() freed the mp. */
+				mp = next;
+				continue;
+			}
 			tmphead = tmp;
 			tmptail = tmp;
 			count++;
